# Retrotransposon-insert-validation

A set of snakemake rules and scripts for validating the calls generated by the [Retrotransposon Insertion Detection Pipeline](https://github.com/adcosta17/Retrotransposon-insert-detection)

## Dependencies
- pysam
- bgzip
- snakemake
- minimap2
- winnowmap2
- lastal
- samtools
- longshot
- whatshap

## Overview

This repo contains a set of snakemake rules and scripts that are used to generate the data used in, and evaluate the results of experiments designed to assess the performance of the [Retrotransposon Insertion Detection Pipeline](https://github.com/adcosta17/Retrotransposon-insert-detection).

### Deletion Based Validation

A deletion based validation approach is used to assess the precision and recall of the pipeline using PacBio Hifi based assemblies and Nanopore long reads.
We downloaded the PacBio Hifi based diploid assemblies and Nanopore reads from the [HPRC](https://github.com/human-pangenomics/HPP_Year1_Data_Freeze_v1.0) for samples: [HG00438](https://s3-us-west-2.amazonaws.com/human-pangenomics/index.html?prefix=working/HPRC/HG00438/), [HG00621](https://s3-us-west-2.amazonaws.com/human-pangenomics/index.html?prefix=working/HPRC/HG00621/), [HG00673](https://s3-us-west-2.amazonaws.com/human-pangenomics/index.html?prefix=working/HPRC/HG00673/), [HG00735](https://s3-us-west-2.amazonaws.com/human-pangenomics/index.html?prefix=working/HPRC/HG00735/), and [HG00741](https://s3-us-west-2.amazonaws.com/human-pangenomics/index.html?prefix=working/HPRC/HG00741/).
Please see the methods section of the manuscript for a full breakdown of the validation methods. 

#### Prerequisites & Setup

To generate modified fastqs the snakemake pipeline requires a folder for each sample. Within each sample folder a subfolder called `fastq` that contains the original fastqs is also required. The fastqs should be compressed using bgzip for pysam to be able to read them. Additionally a second subfolder called `assembly` should be generated that contains the maternal and paternal contigs downloaded from the HPRC. The maternal should be named `<sample>.mat.fa` and the paternal `<sample>.pat.fa`. 

Each HPRC sample listed contains three fastqs each. These can be downloaded directly to the sample's fastq folder. We can then generate deletion fastqs by listing each sample in the config and running the snakemake listed below:

```sh
# Generate Deletion fastqs for HPRC Samples
snakemake -s validation.smk --configfile project_config.yaml all_deletion_fastq
```

#### Results


Once the deletion fastqs have been generated, the main retrotransposon insertion detection pipeline can be run on the modified fastqs. The results of this pipeline can be compared to the annotation file generated when the deletion fastqs are created.

```sh
# Get TP, FP, and FN values for Other HPRC Samples deletion based validation runs
python scripts/get_combined_metrics_deletions.py --input-prefix <path/to/rt_insert_pipeline_run_folder> --id-list <samples,csv,list> --folder winnow_realign_read_analysis --bam-folder winnow_realign
```








